File: dynamic-programming/maximal-square/maximal-square.java
class Solution {
  public int maximalSquare(char[][] matrix) {
    int h = matrix.length;
    int l = matrix[0].length;

    int[][] dp = new int[h][l];

    int ans = 0;

    for (int i = 0; i < h; i++) {
      for (int j = 0; j < l; j++) {
        int curr = matrix[i][j];

        if (curr == '1') {
          dp[i][j] = 1;

          if (i > 0 && j > 0) {
            dp[i][j] += Math.min(Math.min(dp[i - 1][j], dp[i - 1][j - 1]), dp[i][j - 1]);
          }

          ans = Math.max(dp[i][j], ans);
        }
      }
    }
    return ans * ans;
  }
}

File: dynamic-programming/maximal-square/maximal-square.ts
function maximalSquare(matrix: string[][]): number {
  const h: number = matrix.length;
  const l: number = matrix[0].length;

  const dp: number[][] = new Array(h).fill(undefined).map(() => new Array(l).fill(0));

  let ans: number = 0;

  for (let i = 0; i < h; i++) {
    for (let j = 0; j < l; j++) {
      if (matrix[i][j] === '1') {
        dp[i][j] = 1;
        if (i > 0 && j > 0) {
          dp[i][j] += Math.min(Math.min(dp[i][j - 1], dp[i - 1][j]), dp[i - 1][j - 1]);
        }

        ans = Math.max(dp[i][j], ans);
      }
    }
  }
  return ans * ans;
}
